File: laser_war/progress.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from .storage import write_json_atomic

PROGRESS_VERSION = 1


@dataclass
class Progress:
    ultra_unlocked: bool = False

    @classmethod
    def load(cls, path: Path) -> Progress:
        """Load validated unlock progress or return a locked default."""
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, ValueError, TypeError):
            return cls()
        if data.get("version") != PROGRESS_VERSION:
            return cls()
        return cls(ultra_unlocked=bool(data.get("ultra_unlocked")))

    def save(self, path: Path) -> None:
        """Persist unlock progress atomically."""
        write_json_atomic(
            path,
            {
                "version": PROGRESS_VERSION,
                "ultra_unlocked": self.ultra_unlocked,
            },
        )

    def record_result(self, *, mode: str, difficulty: str, winner: str | None) -> bool:
        """Unlock Ultra after the first qualifying Hard-mode victory."""
        if self.ultra_unlocked or mode != "computer" or difficulty != "hard" or winner != "bottom":
            return False
        self.ultra_unlocked = True
        return True

File: web/progress.js
export const PROGRESS_KEY = "laser-war.progress.v1";
export const PROGRESS_VERSION = 1;

/** Load validated Ultra progress or return the locked default. */
export function loadProgress(storage) {
  try {
    const data = JSON.parse(storage.getItem(PROGRESS_KEY));
    if (data?.ultraUnlocked === true) {
      return { ultraUnlocked: true };
    }
    if (data?.version === PROGRESS_VERSION) {
      return { ultraUnlocked: Boolean(data.ultraUnlocked) };
    }
  } catch {
    // Invalid progression data falls back to a locked Ultra mode.
  }
  return { ultraUnlocked: false };
}

/** Persist normalized Ultra progress in browser storage. */
export function saveProgress(storage, progress) {
  storage.setItem(PROGRESS_KEY, JSON.stringify({
    version: PROGRESS_VERSION,
    ultraUnlocked: Boolean(progress.ultraUnlocked),
  }));
}

/** Recover an earned Ultra unlock from a legacy active Ultra match. */
export function recoverUltraProgress(progress, activeSave) {
  if (progress.ultraUnlocked || activeSave?.difficulty !== "ultra") return false;
  progress.ultraUnlocked = true;
  return true;
}

/** Unlock Ultra after the first qualifying Hard-mode victory on either side. */
export function recordResult(progress, {
  mode,
  difficulty,
  winner,
  humanSide = "bottom",
}) {
  if (
    progress.ultraUnlocked
    || mode !== "computer"
    || difficulty !== "hard"
    || winner !== humanSide
  ) {
    return false;
  }
  progress.ultraUnlocked = true;
  return true;
}
